Add pen tool for freehand drawing in PhysicsJS activity

Shashank-Rajanal opened the pull request:
This PR adds a Pen tool to the PhysicsJS activity that allows users to draw freehand shapes directly on the canvas.

Changes included:
- Added a Pen button to the top toolbar
- Implemented freehand drawing using the existing interaction system
- Added a pen icon for the new tool

A short demo video is attached for easier review. Please let me know if any changes are needed.
https://github.com/user-attachments/assets/b858c339-aa4c-4d71-99c2-5889a8c8a98c

ShashankRajanal7 committed:
Add pen tool for freehand drawing in PhysicsJS activity

llaske commented on activities/PhysicsJS.activity/icons/pen.png:
Icon is suppose to be SVG file.
I suggest to reuse this one https://github.com/llaske/sugarizer/blob/dev/activities/Paint.activity/icons/tool-pencil.svg

llaske commented on activities/PhysicsJS.activity/index.html:
```diff
@@ -29,6 +29,7 @@
   <button class="toolbutton" id="box-button" title="Box"></button>
   <button class="toolbutton" id="triangle-button" title="Triangle"></button>
   <button class="toolbutton" id="polygon-button" title="Polygon"></button>
+  <button class="toolbutton" id="pen-button" title="Draw Shape"style="background-image: url(icons/pen.png);"></button>
```
The style is suppose to be in `activity.css`. Have a look on other icons.

llaske commented on activities/PhysicsJS.activity/js/activity.js:
The whole file is change because your editor has replaced TAB indentation by SPACE.
PR should contains only lines updated/added.

llaske reviewed (commented):
Nice start. See comments.

Other remarks:
* [x] When I don't close the polygon, the figure stay in the air and there is a console error. Will be better to just clean the figure.

<img width="1148" height="529" alt="image" src="https://github.com/user-attachments/assets/6a2fbc44-5f2f-485a-9dd6-e3d90f8b7aec" />

* [x] The save in the Journal doesn't work. When the activity is reopen polygons disappeared.

Shashank-Rajanal commented:
@llaske I have resolved all the issues/suggestions which are mentioned in the comments and also worked on "saving the drawings drawn using pen in the journal issue".

By the way , when i run it in the local environment using electron.. even if i don't close the polygon, it creates the shape and everything works fine as in the video below:

https://github.com/user-attachments/assets/4d40ceb1-de12-4a80-9569-5c08fa951e92

ShashankRajanal7 committed:
Add pen tool with minimal changes (no formatting noise)

ShashankRajanal7 committed:
Removed the //modified and //closed comment.

llaske commented:
The problem with not closed polygon is not systematic. See the video below how to try to reproduce the issue.

https://github.com/user-attachments/assets/20b43fe0-6952-449f-805b-327695e80cd7

Two other points:
- I'm not a big fan of your implementation. You are using a specific variable `penMode` to handle the drawing mode but the `currentType` variable is specifically dedicated to handle the current type of volume. So please, try to match to the current philosophy.
- You're not supposed to send a PR on the master branch. See https://github.com/llaske/sugarizer/blob/master/CONTRIBUTING.md

ShashankRajanal7 committed:
Fix polygon closing issue and refractored instead of penMode

Shashank-Rajanal commented:
Hi @llaske i have pushed the updates addressing your feedback.

1. I have sent the PR on the dev branch.
2. Removed 'penMode' variable - Now using 'currentType = 4' to match the existing philosophy .

Also in my local dev environment, polygon with not closed polygon are also generating the shapes as you can see in the video, i tried the same with my friend's dev environment, even there it's working fine.
https://github.com/user-attachments/assets/96d0f59a-5cd2-4230-b799-968e3d543e34

Please let me know if you'd like any further adjustments or if you're still seeing issues on your end.
Thanks for your guidance!

llaske commented on activities/PhysicsJS.activity/js/activity.js:
```diff
@@ -118,20 +133,32 @@ define(["sugar-web/activity/activity","tutorial","l10n","sugar-web/env"], functi
 				switchToType(currentType);
+				isDrawing = false;
+				drawingPoints = [];
+				clearTempCanvas();
```
Avoid repeat these 3 lines on each type. Move them in switchToType function.

llaske commented on activities/PhysicsJS.activity/icons/pen.png:
This file should be removed now

llaske commented:
Nice.
See my comments on your source code.
Also you should add a step on tutorial and string localized.

ShashankRajanal7 committed:
Added pen tool tutorial

Shashank-Rajanal committed:
Merge branch 'dev' into physicsjs-pen-tool

Shashank-Rajanal commented:
@llaske I have added a step on tutorial for the pen tool and is working fine.
Is there any issues in the source code? because you are asking to "see comments on your source code", i didn't get that.. can u be specific?

llaske reviewed (changes requested):
Sorry, forgot to send code change

ShashankRajanal7 committed:
Removed three lines of code from each button handlers and added to switchToType function

Shashank-Rajanal commented:
Done, i have removed those lines from each button handlers and then added it to the switchToType function, then removed pen.png.

llaske commented:
Two remarks:

- On iOS/Android there is a shift with the starting point of the draw (see the initial vertical bar on the video)

https://github.com/user-attachments/assets/dd9af3db-6087-4ded-9e60-cc8445925704

- Provide the strings in all language - you can use english for all

nithinshiju-dev commented:
@llaske can i work on this?

ShashankRajanal7 committed:
offset change issue resolved

Shashank-Rajanal commented:
@nithinshiju-dev , I have already worked on this.

ShashankRajanal7 committed:
Provided strings in all languages

Shashank-Rajanal commented:
@llaske, I have resolved the offset issue and also provided the strings  in all the languages, check the work and let me know if there are any issues.

nithinshiju-dev commented:
> @nithinshiju-dev , I have already worked on this.

Ohh sorry!

llaske commented:
There is lot error in console during activity. Is it possible to fix it?

<img width="1147" height="678" alt="image" src="https://github.com/user-attachments/assets/4126269d-8d52-4b78-a004-c5465349c4e5" />

Shashank-Rajanal commented:
@llaske I have worked on the console errors, it should be good now. Do check it out and let me know if there's any issue.

ShashankRajanal7 committed:
console errors: fixed

llaske commented:
Still the same on my side.

<img width="910" height="650" alt="image" src="https://github.com/user-attachments/assets/c059aa22-ba52-433b-9613-ae31345fed20" />

ShashankRajanal7 committed:
Console errors fixed

Shashank-Rajanal commented:
@llaske, i have edited the code, check if the console errors are still arising or is it resolved.

llaske commented:
Still the same on my side.

<img width="1089" height="669" alt="image" src="https://github.com/user-attachments/assets/a5827c03-556a-4bf8-a660-126517023e22" />

Did you test it?

Shashank-Rajanal commented:
since I'm using electron dependency, it opens an app where i cant open the console. So i was not able to test it.
 Can u tell me how to run the file locally through web interface, so that i can open the console and resolve it parallelly?

HarleenKaurBhullar commented:
Hello I stumbled upon your PR, and just wanted to help in electron you can open using npm start -- --sdebug it worked for me.

nithinshiju-dev commented:
You can also run it via the web interface. Clone sugarizer-server from
https://github.com/llaske/sugarizer-server

then go into that directory and run npm start. Open it in the browser and you’ll be able to use the console for debugging.

llaske commented:
> since I'm using electron dependency, it opens an app where i cant open the console. So i was not able to test it. Can u tell me how to run the file locally through web interface, so that i can open the console and resolve it parallelly?

you should run electron with the sdebug option.
for example:

```npm start -- --window --sdebug```

llaske commented:
@Shashank-Rajanal are you still working on it?

nithinshiju-dev commented:
@llaske can i try working on this?

Shashank-Rajanal commented:
@llaske I tried to resolve it but couldn't find the bug, it will be better if someone volunteers and help it resolve.

HarleenKaurBhullar commented:
@Shashank-Rajanal I wanted to help you but when i ran your pr locally no errors appeared on my side, is there something specific causing these errors.

llaske commented:
Everyone is welcome to try to fix the issue.
The first step is to solve conflicts because the Physics activity has changed since this PR has been raised.

llaske commented:
Fixed in https://github.com/llaske/sugarizer/pull/1985
